src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+use Traversable;
+use function array_column;
+use function array_map;
+use function array_pop;
+use function count;
+use function rewind;
+use function sprintf;
+
+final class DoctrineItems implements IteratorAggregate, Items
+{
+    private $connection;
+    private $tableItems;
+    private $tableVersions;
+
+    public function __construct(Connection $connection, string $tablePrefix)
+    {
+        $this->connection = $connection;
+        $this->tableItems = "{$tablePrefix}_items";
+        $this->tableVersions = "{$tablePrefix}_versions";
+    }
+
+    public function add(ItemVersion $item) : void
+    {
+        $this->connection->transactional(
+            function () use ($item) : void {
+                try {
+                    $current = $this->getCurrentVersion($item->getId(), LockMode::PESSIMISTIC_WRITE);
+                } catch (ItemNotFound $e) {
+                    if (1 !== $item->getVersion()->toInt()) {
+                        throw new UnexpectedVersionNumber(
+                            $item->getId(),
+                            $item->getVersion(),
+                            ItemVersionNumber::fromInt(1)
+                        );
+                    }
+
+                    $this->insertNewItem($item);
+
+                    return;
+                }
+
+                $next = $current->next();
+
+                if ($item->getVersion() > $next) {
+                    throw new UnexpectedVersionNumber($item->getId(), $item->getVersion(), $next);
+                }
+
+                if ($item->getVersion() < $next) {
+                    $this->updateVersion($item);
+
+                    return;
+                }
+
+                $this->insertVersion($item);
+            }
+        );
+    }
+
+    public function remove(ItemId $id, ?ItemVersionNumber $version) : void
+    {
+        $this->connection->transactional(
+            function () use ($id, $version) : void {
+                if (null === $version) {
+                    $this->deleteItem($id);
+
+                    return;
+                }
+
+                try {
+                    $current = $this->getCurrentVersion($id, LockMode::PESSIMISTIC_WRITE);
+                } catch (ItemNotFound $e) {
+                    return;
+                }
+
+                if ($version > $current) {
+                    return;
+                }
+
+                if ($version < $current) {
+                    throw new UnexpectedVersionNumber($id, $version, $current);
+                }
+
+                if (1 === $current->toInt()) {
+                    $this->deleteItem($id);
+
+                    return;
+                }
+
+                $this->deleteVersion($id, $current);
+            }
+        );
+    }
+
+    public function get(ItemId $id, ?ItemVersionNumber $version = null) : ItemVersion
+    {
+        if (null === $version) {
+            $sql = 'SELECT version.id, version.version, version.content, version.hash'
```

Comment:
`version.*` to be more abstract? Other heavy columns around?

Reply:
They're the only columns. Just favouring being explicit.